### ISSUE TITLE
prompt toolkit spawns too many threads

### ISSUE THREAD

**Issue by amigrave** [prompt-toolkit/python-prompt-toolkit#881]
Every type stroke in a prompt toolkit spawns some threads, when typing fast there can be a huge amount of threads spawned concurrently. Constrained environments such as containers with some cgroups limits on the amount of processes prompt toolkit can crash when typing fast.

Here's a live demo that showcase the threads usage:
https://asciinema.org/a/GrtgrvnaGxUszAVXQpnp2KSYI

**Comment by jonathanslenders** [comment]
Hi @amigrave,

I expect that this is coming from certain timeouts that are triggered on key strokes.

It is not so easy to solve, but right now, I'm working on the upgrade to Python 3 only, which will use asyncio for the event loop. As I understand, asyncio has a much better timer implementation, and won't require threads for each timer.
I can't promise a release date yet, it's quite a bit of work, but I'm eager to complete the upgrade.

**Comment by amigrave** [comment]
@jonathanslenders Thanks for your answer! Just out of curiosity, are you working in a specific branch for this python3 only upgrade?

**Comment by jonathanslenders** [comment]
Hi @amigrave,
Check the master branch right now. This contains the latest version I've been working on.
It uses asyncio natively, and I expect fewer threads. Let me know how it goes.
IPython/ptpython/etc... need little updates to work with prompt_toolkit 3.0, so I don't think you can test with those out of the box.

**Comment by amigrave** [comment]
@jonathanslenders Fantastic!  I did the test again comparing the master branch and a checkout at revision eee625b, with commit eee625b I could experience the same behavior I explained, and with master (actually at revision 7b16d24) the single process/single thread nature of your new asyncio implementation fixed the issue.

Again, out of curiosity, how hard would you describe the conversion from threading to asyncio you made on prompt-toolkit? Looking at the commit c7851c0 it's hard to tell because it seems you did the asyncio conversion, added type annotation and made some pep8 work in the same commit :)

**Comment by jonathanslenders** [comment]
Hi @amigrave,
For most applications, it would be trivial to upgrade to prompt_toolkit 3, once that is released. I will test with a few applications and then write down the required instructions. It's likely that many applications don't need to do any changes at all. IPython requires a few tests, because the way inputhooks are set up is slightly different. But similar enough that's easy to support both prompt_toolkit 2 and 3 at the same time.

I squashed these changes in one commit, because I had many experimental changes, all doing a bit of both, and didn't want to bother splitting it up in logical commits.
But the async changes in prompt_toolkit basically consist of:
- Removing most of the custom eventloop code that we had + the implementation of coroutines, futures and async generators.
- Replacing all async functions to use the async/await syntax.
- Use of real async generators.
- The input hook code changed (mostly important for IPython). While previous, the prompt_toolkit event loops took an input hook as input, it now became part of a selector for an asyncio eventloop.
- The dialogs API was changed a little.
